include/exadg/operators/operator_base.cpp

```diff
@@ -2153,6 +2155,173 @@ OperatorBase<dim, Number, n_components>::evaluate_face_integrals() const
          (integrator_flags.face_integrate != dealii::EvaluationFlags::nothing);
 }
 
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::compute_factorized_as_matrices() const
+{
+#ifdef DEAL_II_WITH_TRILINOS
+  internal_compute_factorized_as_matrices<dealii::TrilinosWrappers::SparseMatrix>();
+#elif DEAL_II_WITH_PETSC
+  internal_compute_factorized_as_matrices<dealii::PETScWrappers::MPI::SparseMatrix>();
+#else
+  AssertThrow(n_mpi_processes == 1,
+              ExcMessage(
+                "Use Trilinos or Petsc for distributed sparse matrices needed for this function!"));
+  internal_compute_factorized_as_matrices<dealii::SparseMatrix>();
+#endif
+}
+
+template<int dim, typename Number, int n_components>
+template<typename SparseMatrix>
+void
+OperatorBase<dim, Number, n_components>::internal_compute_factorized_as_matrices() const
+{
+  dealii::DoFHandler<dim> const & dof_handler =
+    this->matrix_free->get_dof_handler(this->data.dof_index);
+
+  unsigned int const dofs_per_cell = matrix_free->get_dofs_per_cell(this->data.dof_index);
+
+  unsigned int const n_cells = matrix_free->n_cell_batches() * vectorization_length;
+
+  // assemble a temporary sparse matrix to cut out the blocks
+  SparseMatrix                   tmp_matrix;
+  dealii::DynamicSparsityPattern dsp;
+  internal_init_system_matrix(tmp_matrix, dsp, dof_handler.get_communicator());
+  internal_calculate_system_matrix(tmp_matrix);
+
+  // cut out overlapped block matrices
+  std::vector<dealii::FullMatrix<Number>> overlapped_block_matrix(
+    n_cells, dealii::FullMatrix<Number>(dofs_per_cell));
+
+  std::vector<std::vector<dealii::types::global_dof_index>> all_dof_indices(
+    n_cells, std::vector<dealii::types::global_dof_index>(dofs_per_cell));
+
+  initialize_dof_vector(weights);
+  for(unsigned int cell_batch_index = 0; cell_batch_index < matrix_free->n_cell_batches();
+      ++cell_batch_index)
+  {
+    unsigned int const n_filled_lanes =
+      matrix_free->n_active_entries_per_cell_batch(cell_batch_index);
+
+    for(unsigned int v = 0; v < n_filled_lanes; ++v)
+    {
+      auto const & cell = matrix_free->get_cell_iterator(cell_batch_index, v);
+
+      auto & local_dof_indices = all_dof_indices[cell_batch_index * vectorization_length + v];
+      if(is_mg)
+        cell->get_mg_dof_indices(local_dof_indices);
+      else
+        cell->get_dof_indices(local_dof_indices);
+
+      for(auto const & i : local_dof_indices)
+        weights[i] += 1.;
+    }
+  }
+  weights.compress(dealii::VectorOperation::add);
+
+  for(unsigned int i = 0; i < weights.size(); ++i)
+  {
+    if(weights.in_local_range(i))
+      weights[i] = 1. / std::sqrt(weights[i]);
+  }
+  weights.update_ghost_values();
```

Comment:
Could you explain via comments the idea behind the "weights" calculations?

Reply:
In particular, where does `1./std::sqrt()` come from?

Reply:
The idea is to weight contributions which have overlap with `1/n_{contributions to dof}`. To do this in a symmetric way we weight the local src vector and the local dst vector with `1./std::sqrt()`. I will add a comment!

Reply:
I added comments in 51c103e259e52b893757735d59929de446234d56